Use trees in selects when hierarchy is present

ptbrowne opened the pull request:
## How to test

- Use bathing water with hierarchies (check draft when searching)
- In the data filters, the "observation location" select should display the hierarchy as a tree
- When activating interactive filters, the interactive filter also displays the hierarchy as a tree

## TODO

- [x] When you first open the select, it loads the hierarchy and then closes the select, and you have to open it again

bprusinowski commented on app/components/select-tree.tsx:
```diff
@@ -0,0 +1,351 @@
+  );
+
+  const menuRef = React.useRef<PopoverActions>(null);
+  const handleNodeToggle = () => {
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 500);
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 1000);
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 2000);
+  };
```
Why such combo? Wouldn't just first timeout do the job? Or is the transition duration dependent on the number of children in a node?

ptbrowne replied:
Refactored away by using the onEntered / onExited transition props of the tree item transitions.

bprusinowski commented on app/components/select-tree.tsx:
```diff
@@ -0,0 +1,351 @@
+          onNodeToggle={handleNodeToggle}
+          defaultCollapseIcon={<ExpandMoreIcon />}
+          defaultExpandIcon={<ChevronRightIcon />}
+          // onNodeSelect={handleNodeSelect}
```
👁️

bprusinowski commented on app/docs/form.docs.tsx:
```diff
@@ -28,6 +29,49 @@ const SwitchExample = ({ initialChecked }: { initialChecked?: boolean }) => {
+    value: `example_${i}`,
+    label: `Example ${i}`,
+  }));
+
+const SelectTreeExample = () => {
+  const [value, setValue] = useState("5");
+  return (
+    <SelectTree
+      value={value}
+      onChange={({ target: { value } }) => setValue(value)}
+      options={[
+        {
+          value: "1",
+          label: "Applications",
+          children: [{ value: "2", label: "Calendar" }],
+        },
+        {
+          value: "5",
+          label: "Documents",
+          selectable: false,
+          children: [
+            { value: "10", label: "OSS" },
+            {
+              value: "6",
+              label: "MUI",
+              children: [{ value: "8", label: "index.js" }],
+            },
+          ],
+        },
+        {
+          label: "Numbers",
+          value: "numbers",
+          selectable: false,
+          children: numberOptions,
+        },
+      ]}
+    />
+  );
+};
+
```
👌

bprusinowski commented on app/charts/shared/chart-data-filters.tsx:
```diff
@@ -255,14 +269,63 @@ const DataFilterBaseDimension = ({
+    description: tooltipText,
+    values: dimensionValues,
+  } = dimension;
+  const options = React.useMemo(() => {
+    let opts = [] as { label: string; value: string; isNoneValue?: boolean }[];
+    if (hierarchy) {
+      opts = hierarchyToOptions(hierarchy);
+    } else {
+      opts = dimensionValues;
+    }
+    if (!isKeyDimension) {
+      opts.push({
+        value: FIELD_VALUE_NONE,
+        label: noneLabel,
+        isNoneValue: true,
+      });
+    }
+    return opts;
+  }, [hierarchy, isKeyDimension, dimensionValues, noneLabel]);
```
Did not check this as I'm not aware of optional hierarchy in any of the datasets, but wouldn't this append the "None" option at the end of the array (I think it should be at top)?

ptbrowne replied:
Ah yes ! I replaced push by unshift :)

ptbrowne commented:
1. Yes I need to see how to still lazy load the hierarchies (we had gained a significant boost by postponing the downloading of hierarchies), but it should be better done, I need to do refactors here to correctly provide the loading to the select.
2. Fixed, the method may seem hacky (via the --depth CSS variable), but I tried first to use calc(--depth) + 1, but it does not seem to work, it seems we can't have recursion in CSS variables :/
3. I tried to add an indication, I think it's better now, what do you think ? (cc @AnninaWalker )
4. Since I changed the way of doing the padding for the hover effect, this has changed I think. I personally think this looks good already ? 

<img width="318" alt="image" src="https://user-images.githubusercontent.com/465582/202484116-76ad5ea9-3034-4994-a60b-88847f9dc013.png">

bprusinowski reviewed (approved):
Yes, much better now 💯
